Add access_services.py to access services locally

sbernauer opened the pull request:
## Description
Intended for developing and demos.
Choosing port-forward over collecting Node-IPs and Node-Ports and constructing URLs as this also works if you may don't have access to the Nodes directly(?). (I'm used to run stuff on a managed k8s and not on my machine)
Choose python over bash as it matches [create_test_cluster.py](https://github.com/stackabletech/integration-tests/blob/main/create_test_cluster.py) and it got a little bit complicated for bash in my opinion.

Demo usage with all the operators i could come up with
```
sbernauer@sbernauer:~/stackabletech/integration-tests$ ./forward_services_locally.py 
+-------------+------------------------------------------+--------+------------+------------------------+
| Namespace   | Service                                  |   Port | Name       | URL                    |
|-------------+------------------------------------------+--------+------------+------------------------|
| default     | airflow-webserver                        |   8080 | airflow    | http://localhost:33897 |
| default     | airflow-webserver-default                |   8080 | airflow    | http://localhost:47223 |
| default     | minio                                    |     80 | http-minio | http://localhost:33251 |
| default     | simple-derby-druid-broker                |   8082 | http       | http://localhost:54235 |
| default     | simple-derby-druid-broker-default        |   8082 | http       | http://localhost:53329 |
| default     | simple-derby-druid-coordinator           |   8081 | http       | http://localhost:47055 |
| default     | simple-derby-druid-coordinator-default   |   8081 | http       | http://localhost:41185 |
| default     | simple-derby-druid-historical            |   8083 | http       | http://localhost:56459 |
| default     | simple-derby-druid-historical-default    |   8083 | http       | http://localhost:59267 |
| default     | simple-derby-druid-middlemanager         |   8091 | http       | http://localhost:55585 |
| default     | simple-derby-druid-middlemanager-default |   8091 | http       | http://localhost:56739 |
| default     | simple-derby-druid-router                |   8888 | http       | http://localhost:57035 |
| default     | simple-derby-druid-router-default        |   8888 | http       | http://localhost:54603 |
| default     | simple-hbase-master-default              |  16010 | ui         | http://localhost:49947 |
| default     | simple-hbase-regionserver-default        |  16030 | ui         | http://localhost:39591 |
| default     | simple-hdfs-datanode-default             |   9864 | http       | http://localhost:57713 |
| default     | simple-hdfs-journalnode-default          |   8480 | http       | http://localhost:56443 |
| default     | simple-hdfs-journalnode-default          |   8481 | https      | http://localhost:39829 |
| default     | simple-hdfs-namenode-default             |   9870 | http       | http://localhost:52317 |
| default     | simple-nifi                              |   8443 | https      | http://localhost:58427 |
| default     | simple-nifi-node-default                 |   8443 | https      | http://localhost:53229 |
| default     | simple-opa                               |   8081 | http       | http://localhost:48709 |
| default     | simple-opa-server-default                |   8081 | http       | http://localhost:55137 |
| default     | simple-spark-history-server-default      |  18080 | http       | http://localhost:45583 |
| default     | simple-spark-master                      |   8080 | http       | http://localhost:37117 |
| default     | simple-spark-master-default              |   8080 | http       | http://localhost:39787 |
| default     | simple-spark-slave-default               |   8081 | http       | http://localhost:38053 |
| default     | simple-superset-external                 |   8088 | superset   | http://localhost:34111 |
| default     | simple-superset-node-default             |   8088 | superset   | http://localhost:35505 |
| default     | simple-trino-coordinator                 |   8080 | http       | http://localhost:60059 |
| default     | simple-trino-coordinator-default         |   8080 | http       | http://localhost:50231 |
| default     | simple-trino-worker-default              |   8080 | http       | http://localhost:43043 |
+-------------+------------------------------------------+--------+------------+------------------------+

```

## Review Checklist
- [ ] Code contains useful comments
- [ ] Changelog updated (or not applicable)

maltesander reviewed (commented):
Nice i like it. Some comments:
1) I would make the namespace a cli argument (which can default to "default")
2) I had to install "kubernetes" and "tabulate" libs via pip, so maybe add a requirements.txt?
3) Do we want to pick this up in the readme? 
4) Druid and HBase come to mind for additional services to expose?

sbernauer commented:
Many thanks for your feedback @maltesander! I refactored the script quite a bit
* Added some CLI args
* Added all operators i could come up with
* I thought about it and i not wanted to restrict the services only to the "load-balanced". Instead i included the roleGroup services by now (some products dont have a load-balanced service). We can easily change that.
* As now the list exceeded 100 forwards i filtered down by the name of the port
* Excluded metrics by now. We can add a CLI argument for that if we want.
* Also print the Port name and not only the Port number :)

sbernauer commented on forward_services_locally.py:
```diff
@@ -0,0 +1,111 @@
+    r".*-spark-master$",
+    r".*-spark-history-server",
+    r".*airflow-webserver$",
+]
```
Thanks for the suggestion! I actually changed the service List/Map significant, please have a look at the new version

maltesander reviewed (commented):
Awesome thank you. I have one more future idea to use the specified service container ports instead of hardcoding the port names (which may change), but happy to leave that to future development. Any reason you removed zookeeper and kafka?

maltesander reviewed (dismissed):
LGTM if you wanna merge!

sbernauer commented:
>  I have one more future idea to use the specified service container ports instead of hardcoding the port names (which may change)

I first implemented it this way but as i saw the many port numbers i though it's easier to maintain the names. I also think they names should be more stable than the actual numbers but I'm not sure.

> Ah i would filter for the service name to be honest. Airflows "airflow" port might be on top and a "rpc" "http" is something to search for further down?

I'm not exactly getting this, as we already filter by the service name (with the regex). Do you want to include all the "not a webinterface" ports and put them in a list below? I had > 120 service ports running which results in a very long list. And we have a flag `--all-services` to do that

maltesander commented:
> I first implemented it this way but as i saw the many port numbers i though it's easier to maintain the names. I also think they names should be more stable than the actual numbers but I'm not sure.

Ok when i check the list above, we still have inconsistencies (e.g. hbase ui <--> http). I think reading service ports is more stable than hard coding (if new operators come, we enable https for tools etc.) 

> I'm not exactly getting this, as we already filter by the service name (with the regex). Do you want to include all the "not a webinterface" ports and put them in a list below? I had > 120 service ports running which results in a very long list. And we have a flag `--all-services` to do that

Yeah i think i misunderstood. I thought you sort purely for the port name but it is not the case. You can ignore this.

soenkeliebau commented:
That looks awesome!
Just one (a bit fundamental, sorry) comment. The `./get_url.sh` script didn't rely on port forwarding to create connectivity, but looked up the external ip addresses of nodes running the pods and then used the services to generate links that work without forwarding any ports.
This will of course not work in all network scenarios, but if it does work, I think it should be preferred over port forwarding. Could we maybe add a switch to the script to pick which method should be used?

sbernauer commented:
@maltesander if it is okay for you i would proceed with the port names for now - happy to change that later on. I honestly still think port numbers are the least stable thing in a service ^^
@soenkeliebau i will look into not using port-forward every time - the original motivation for using port-forward is in the PRs description

soenkeliebau commented:
@sbernauer contacted me about getting the script, which I forgot to include in my comment, I'll add it here for reference.


```
#!/usr/bin/env bash

DRUID_URL=`kubectl get svc demo-druid-router -o json | jq -r --argfile endpoints <(kubectl get endpoints demo-druid-router -o json) --argfile nodes <(kubectl get nodes -o json) '($nodes.items[] | select(.metadata.name == $endpoints.subsets[].addresses[].nodeName) | .status.addresses | map(select(.type == "ExternalIP" or .type == "InternalIP")) | min_by(.type) | .address | tostring) + ":" + (.spec.ports[] | select(.name == "http") | .nodePort | tostring)'`
DRUID_RC=$?

echo "Druid UI: http://${DRUID_URL}"
```

> Choosing port-forward over collecting Node-IPs and Node-Ports and constructing URLs as this also works if you may don't have access to the Nodes directly(?). (I'm used to run stuff on a managed k8s and not on my machine)

I have used the ./get_url script with both managed k8s on Google Cloud and managed kubernetes from Ionos and it worked quite well for both of those. On Google I had to add a firewall rule to allow traffic in, on Ionos it just worked.

My motivation for going this route was that NiFi for _reasons_ doesn't listen on the local interface and so sadly doesn't work with port forwards.

maltesander commented:
> @maltesander if it is okay for you i would proceed with the port names for now - happy to change that later on. I honestly still think port numbers are the least stable thing in a service ^^ 

I think you got me wrong, i didnt mean to fix the port numbers, i just thought listing all the service ports for a service and take em up here. I guess the rpc ports only make sense for sth. like spark master, but that would require no hardcoding and is automatically adapting to port name or number changes?

But as i said i think it is already awesome and fine with merging (maybe wait for a go from @soenkeliebau and the non forwarding)

sbernauer commented:
@maltesander to be honest i think i need a short sync to get what you are meaning. But also fine to do that at a later time/leave as is, as you prefer.
@soenkeliebau i ported your shell-script. Sadly i see problems with services not having a nodePort. Do you have any idea how to resolve that? Superset and Trino links work fine.
```
./access_services.py -d
+-------------+----------------------------------+--------+--------------+------------------------------------------------------+
| Namespace   | Service                          |   Port | Name         | URL                                                  |
|-------------+----------------------------------+--------+--------------+------------------------------------------------------|
| default     | alertmanager-operated            |   9093 | http-web     | http://172.18.0.2:<no nodePort attribute on service> |
| default     | minio1-console                   |   9090 | http-console | http://172.18.0.3:<no nodePort attribute on service> |
| default     | prometheus-operated              |   9090 | http-web     | http://172.18.0.4:<no nodePort attribute on service> |
| default     | prometheus-operator-grafana      |     80 | http-web     | http://172.18.0.3:<no nodePort attribute on service> |
| default     | simple-superset-external         |   8088 | superset     | http://172.18.0.2:32579                              |
| default     | simple-superset-node-default     |   8088 | superset     | http://172.18.0.2:<no nodePort attribute on service> |
| default     | simple-trino-coordinator         |   8080 | http         | http://172.18.0.4:30699                              |
| default     | simple-trino-coordinator-default |   8080 | http         | http://172.18.0.4:<no nodePort attribute on service> |
| default     | simple-trino-worker-default      |   8080 | http         | http://172.18.0.3:<no nodePort attribute on service> |
+-------------+----------------------------------+--------+--------------+------------------------------------------------------+

soenkeliebau commented:
> @soenkeliebau i ported your shell-script. Sadly i see problems with services not having a nodePort. Do you have any idea how to resolve that? Superset and Trino links work fine.

I am just a dumb user of this script, @teozkr did all the heavy lifting, I'll defer to him here :)

sbernauer commented:
I had to manually search for the credentials from the examples. What do you think about printing some additional infos?
```
./access_services.py 
+-------------+----------------------------------+--------+--------------+------------------------+---------------------------------------+
| Namespace   | Service                          |   Port | Name         | URL                    | Extra info                            |
|-------------+----------------------------------+--------+--------------+------------------------+---------------------------------------|
| default     | alertmanager-operated            |   9093 | http-web     | http://localhost:53159 |                                       |
| default     | minio1-console                   |   9090 | http-console | http://localhost:57211 | accesskey: minio, secretkey: minio123 |
| default     | prometheus-operated              |   9090 | http-web     | http://localhost:56621 |                                       |
| default     | prometheus-operator-grafana      |     80 | http-web     | http://localhost:47869 | user: admin, password: prom-operator  |
| default     | simple-superset-external         |   8088 | superset     | http://localhost:54609 | user: admin, password: admin          |
| default     | simple-superset-node-default     |   8088 | superset     | http://localhost:43087 |                                       |
| default     | simple-trino-coordinator         |   8080 | http         | http://localhost:53367 |                                       |
| default     | simple-trino-coordinator-default |   8080 | http         | http://localhost:44421 |                                       |
| default     | simple-trino-worker-default      |   8080 | http         | http://localhost:36145 |                                       |
+-------------+----------------------------------+--------+--------------+------------------------+---------------------------------------+

soenkeliebau commented:
For demo purposes that looks good, not sure how we want to handle this going forward though.
I assume you read the credentials from the CRDs directly?

I'd like to keep your script useble for later scenarios where credentials are hidden in LDAP or a tls cert or something similar, so we can still show all endpoints.

soenkeliebau commented:
Thought that just occurred to me, would this script be better placed in the utils repo?

maltesander commented:
> Thought that just occurred to me, would this script be better placed in the utils repo?

Will we move the kuttl tests to the operators?
And should the kuttl tests be independent meaning they create all their dependencies themselves?

Then i think it should stay wherever `create_test_cluster.py` is located, which then will be just for demos and developing?

sbernauer commented:
> Then i think it should stay wherever create_test_cluster.py is located, which then will be just for demos and developing?

I see it the same way. My thought is to place it to the `create_test_cluster.py`, happy to move them together later on

The script reads the secrets from the k8s secret directly (the CRDs don't contain the passwords). Its implemented in a flexible way. If the retrieval of the credentials fails for whatever reason the script still works, only no credentials will be shown for that service. So the script should be future-proof.

sbernauer commented:
Had to rebase, please feel free to approve again

soenkeliebau commented:
I just ran it against my test cluster and got the following failure, probably down to something stupid I did, the secret `demo-superset-credentials` does indeed not exist, but I think it never has..

```
(⎈ |kind-integration-tests:default)➜  demo ./access_services.py
Error from server (NotFound): secrets "demo-superset-credentials" not found
Stopping 11 port-forwards
Traceback (most recent call last):
  File "/home/sliebau/stackable/demo/./access_services.py", line 177, in <module>
    sys.exit(main())
  File "/home/sliebau/stackable/demo/./access_services.py", line 106, in main
    forward_port(service_namespace, service_name, service_port, service_port_name, args.verbose)
  File "/home/sliebau/stackable/demo/./access_services.py", line 140, in forward_port
    get_extra_info(service_name),
  File "/home/sliebau/stackable/demo/./access_services.py", line 166, in get_extra_info
    return subprocess.check_output(complete_command, shell=True).decode('utf-8')
  File "/usr/lib/python3.10/subprocess.py", line 420, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/usr/lib/python3.10/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command 'SERVICE_NAME=demo-superset-external && kubectl get secret $(echo "$SERVICE_NAME" | sed "s/-external$/-credentials/") --template='user: {{index .data "adminUser.username" | base64decode}}, password: {{index .data "adminUser.password" | base64decode}}'' returned non-zero exit status 1.
(⎈ |kind-integration-tests:default)➜  demo
```

sbernauer commented:
Sorry i forgot to push my changes... Failing shell commands for the "Extra info" column shut not stop the whole script from working.

Ah you didn't match my naming convention of the secrets ^^ The more ideal solution would to get the secretName from the CRD and than look into that secret. But i think that's maybe a little bit complicated for now. I would look at the "Extra info" column as best effort job.

sbernauer commented:
@soenkeliebau actually it wasn't that hard to fix. It should work now ;)

soenkeliebau commented:
Looks good, I just ran it and it produced the expected result :)

I have not actually read the code, I am sorry! Just ran pylint against it and it produced a number of warnings, could you maybe have a quick look at those?
I'll create a PR to add linters to this repo as well, seems we have forgotten that.

```
access_services.py:1:0: C0114: Missing module docstring (missing-module-docstring)
access_services.py:46:0: C0103: Constant name "k8s" doesn't conform to UPPER_CASE naming style (invalid-name)
access_services.py:52:0: C0116: Missing function or method docstring (missing-function-docstring)
access_services.py:68:0: C0116: Missing function or method docstring (missing-function-docstring)
access_services.py:69:4: C0103: Constant name "k8s" doesn't conform to UPPER_CASE naming style (invalid-name)
access_services.py:69:4: C0103: Constant name "k8s_nodes" doesn't conform to UPPER_CASE naming style (invalid-name)
access_services.py:69:4: W0602: Using global for 'k8s_nodes' but no assignment is done (global-variable-not-assigned)
access_services.py:114:0: C0116: Missing function or method docstring (missing-function-docstring)
access_services.py:121:0: C0116: Missing function or method docstring (missing-function-docstring)
access_services.py:122:71: C0103: Variable name "s" doesn't conform to snake_case naming style (invalid-name)
access_services.py:126:0: C0116: Missing function or method docstring (missing-function-docstring)
access_services.py:132:51: R1732: Consider using 'with' for resource-allocating operations (consider-using-with)
access_services.py:132:18: R1732: Consider using 'with' for resource-allocating operations (consider-using-with)
access_services.py:143:0: C0116: Missing function or method docstring (missing-function-docstring)
access_services.py:157:0: C0116: Missing function or method docstring (missing-function-docstring)
access_services.py:162:4: R1705: Unnecessary "else" after "return" (no-else-return)
access_services.py:166:15: W1510: Using subprocess.run without explicitly set `check` is not recommended. (subprocess-run-check)
access_services.py:168:0: C0116: Missing function or method docstring (missing-function-docstring)

```
A thought for later, before we put much more effort into this tool I'd like to at least investigate the effort of porting it to Rust. The create-test-cluster Script does not have any dependencies and just works with plain python, for this the user needs to install dependencies. This may be a personal issue of mine, but I get a rash whenever I touch python lib handling :)

That being said, happy to merge pending linter warnings being investigated.

sbernauer commented:
Ouch ^^
Fixed the warnings beside the "Line too long" warning. Now we have a score of `8.14/10`. For comparison create_test_cluster.py has `-4.47/10` :joy:

soenkeliebau commented:
My version of pylint seems to be stricter :)

```
access_services.py:75:4: W0602: Using global for 'K8S_NODE_IPS' but no assignment is done (global-variable-not-assigned)
access_services.py:147:51: R1732: Consider using 'with' for resource-allocating operations (consider-using-with)
access_services.py:147:18: R1732: Consider using 'with' for resource-allocating operations (consider-using-with)


------------------------------------------------------------------
Your code has been rated at 7.94/10 (previous run: 5.88/10, +2.06)
```

sbernauer commented:
LGTM beside the failing check. Maybe it is related to https://github.com/reviewdog/action-hadolint/issues/54

sbernauer commented:
I managed to remove 2 warnings, 2 are left:
```
access_services.py:75:4: W0603: Using the global statement (global-statement)
access_services.py:148:22: R1732: Consider using 'with' for resource-allocating operations (consider-using-with)
```
But to be honest i would leave them as is if that's ok for you.

soenkeliebau commented:
fine for me, yes.
